### ISSUE TITLE
Using CommandBindings with Actions

### ISSUE THREAD

**Issue by Ruhrpottpatriot** [canton7/Stylet#50]
In my app, prior to using Stylet i used CommandBindings and the 200 or so predefined commands to handle the usual actions like open, save, saveas, etc.

In a window you could define them like this:
```
<Window.CommandBindings>
    <CommandBinding Command="Open" Executed="OpenFile_Executed" />
    <CommandBinding Command="Close" Executed="CloseApplication_Executed" />
    <CommandBinding Command="Save" Executed="Save_Executed" />
    <CommandBinding Command="SaveAs" Executed="SaveAs_Executed"/>
</Window.CommandBindings>
```
which then in turn enabled the commands to be used by other elements with intelli-sense support available.
However, if I want to replace the method from the code behind (e.g. OpenFile_Executed) with an action I get an ArgumentException, that Stylet.Xaml.ActionExtension can't be converted to System.Windows.Input.ExecutedRoutedEventHandler

Now the error is pretty clear and the docs are pretty clear how to bind action, but those actions don't have intellisense and I can't access the `ExecutedRoutedEventArgs` which are passed when I'm just using commands. It would also mean I could use pre-defined keyboard shortcuts like Ctrl+S for things like "Save".

Any way to solve this?

**Comment by canton7** [comment]
I'm afraid I'm away from a computer for a week, but I'll have a look as soon as I get back.

Looking at the error, it looks like I need to add support for CommandBinding to Action: if you look at EventAction.cs you'll see how that works.  I'm afraid intellisense doesn't work with actions. Once support is added you'll be able to access the ExecutedRoutedEventArgs.

(If you submit a working PR, AppVeyor will give you a nuget package you can use until I get back to a computer, if that helps things).

**Comment by Ruhrpottpatriot** [comment]
Support for CommandBiding => Action would be great. As I said: This not only gives us additional information and support for hotkeys. It would also open up for custom commands to be created and used. 

On a sidenote and without knowing too much of WPFs internals: Styles currently don't accept actions, only commands. Shouldn't a conversion from CommandBinding (or more like `ICommand`, since `ComandBindings` is just a collection of `CommandBinding` which is a class that holds information about an `ICommand`) to Action solve the problem for Styles?

**Comment by canton7** [comment]
You can use Actions with hotkeys currently: <KeyBinding Command="{s:Action...}" Key="..."/>. You can also use Actions with any property of type ICommand. The tricky bit comes with using Actions with events which don't follow the (object, EventArgs) pattern, as I need to add code to deal with each case which breaks this convention. I thought I'd caught all of the cases in WPF but apparently not. 

Actions work by returning an object which implements ICommand for command properties, or a delegate for events. That however doesn't seem to be enough for it to work with Styles. Binding and Style have a very close (internal) relationship in WPF.

**Comment by canton7** [comment]
What "additional information" does CommandBinding give you, exactly?

**Comment by canton7** [comment]
Right, the problem is that CommandBinding is a DependencyObject, and Action expects to be used with a DependencyProperty on a DependencyObject. This *is* a bit of a problem, as the Action needs to be able to get the ViewModel from the View, and that sort of information doesn't make its way down into the CommandBinding. I can get it in some cases from the IRootObjectProvider, but that won't respect things like s:View.ActionTarget...

**Comment by canton7** [comment]
Right, that's the best I can do.

**Comment by Ruhrpottpatriot** [comment]
A CommandBinding has the following signature: `private void Command_Executed(object sender, ExecutedRoutedEventArgs e)`

The view is the sender and via the EventArgs you also get the original source of the command. Shouldn't that be enough to get all we need?

**Comment by canton7** [comment]
I would have thought that the action would be invoked on the ActionTarget which is present at the point that the CommandBinding is *defined*, rather than at the control which executes the command? It seems very odd to create a CommandBinding which invokes methods on *different* objects depending on where it's used - that seems very counter-intuitive.

**Comment by Ruhrpottpatriot** [comment]
I thinkg this paragraphs from MSDN explains it very neatly:
> When the Execute or CanExecute method of a RoutedCommand is called, the PreviewExecuted/Executed or the PreviewCanExecute/CanExecute events are raised on the command target. If the command target has a CommandBinding for the command, the appropriate handlers are called. If the command target does not have a CommandBinding for the command, the events are routed through the element tree until an element that has a CommandBinding is found.

> A CommandBinding has limited use with an ICommand that is not a RoutedCommand. This is because a CommandBinding binds the command to the ExecutedRoutedEventHandler and the CanExecuteRoutedEventHandler which listen to the Executed and CanExecute routed events which are raised when the Execute and CanExecute method of the RoutedCommand are called.

And a `RoutedCommand` is explained as:
> Defines a command that implements ICommand and is routed through the element tree.

>The Execute and CanExecute methods on a RoutedCommand do not contain the application logic for the command as is the case with a typical ICommand, but rather, these methods raise events that traverse the element tree looking for an object with a CommandBinding. The event handlers attached to the CommandBinding contain the command logic. 

So by design `RoutedCommands` are executed at some point in the element tree, although not necessarily on the action target
I think they did this to allow nested controls to bubble up their commands to the parent view model.
For example you could create a custom control "Named List Box", which is a text block above a regular list box and the list box is hooked up to a double click via a command. Naturally you can't define a handler for the command inside your custom control so you define it in the parent view via a `CommandBinding`.

I don't know how much this impacts your `Action` design, but you stated that they don't work with Styles and you have to use commands instead. By allowing actions into the `CommandBindings` this problem should be solved, shouldn't it?

**Comment by canton7** [comment]
None of the text you quote is relevant to the discussion at hand, which makes me think maybe you're misunderstanding the issue?

I've already added code to allow CommandBindings with Actions: you can see the commits earlier in this thread. 

In pure WPF, when you define a CommandBinding, you write something like `<CommandBinding Command="..." Executed="CommandExecuted"/>`. `CommandExecuted` is a method in the codebehind. When the Command is executed, WPF goes looking for a suitable CommandBinding. If it finds the one we just defined, it executes the `CommandExecuted` method in the codebehind. It doesn't go searching for a method called `CommandExecuted` starting at the point that the RoutedCommand was raised: it uses the method define in the codebehand of the XAML file we defined the `CommandBinding` in.

The whole point of actions is that they invoke methods on the ViewModel rather than on the codebehind. They don't go searching for a suitable ViewModel: they use the ViewModel corresponding to the View which the action is used in. If you write something like `<CommandBinding Command="..." Executed="{s:Action CommandExecuted"/>` in your view, that should always invoke the `CommandExecuted` method in the ViewModel corresponding to the view, **regardless** of where the RoutedCommand was actually raised! 

Again, if you define a CommandBinding in a parent view, it binds to a method on the parent view's codebehind. 
If you then raise the corresponding RoutedCommand in a child view, it calls that method in the parent's codebehind. It doesn't look for a method on the child's codebehind. Using a CommandBinding with an action should behave the same way. It should not start searching for a `CommandExecuted` method on the child's ViewModel: it should go straight for the parent's ViewModel. 

> For example you could create a custom control "Named List Box", which is a text block above a regular list box and the list box is hooked up to a double click via a command. Naturally you can't define a handler for the command inside your custom control so you define it in the parent view via a CommandBinding.

The *normal* way of doing this in MVVM is to define a custom routed event in the NamedListBox (`public static readonly RoutedEvent DoubleClickEvent = ...`, etc). Raise it when your ListBox is double-clicked. In the parent view you would write `<my:NamedListBox DoubleClick="{s:Action NamedListBoxDoubleClicked}" .../>`. Alternatively you can define a DependencyProperty of type `ICommand` in your `NamedListBox`, and wire that up in the parent view with e.g. `<my:NamedListBox DoubleClickCommand="{s:Action NamedListBoxDoubleClicked}" .../>` (doing this gives you access to CanExecute and CanExecuteChanged, using guard properties, as described in the docs.

Note that this is very similar to how Microsoft does things in WPF. A Button defines a RoutedEvent called "Click", and a "Command" DependencyProperty which has type ICommand. Button doesn't just raise a RoutedCommand when it's clicked.

**Comment by azabluda** [comment]
Looking forward to seeing 2027fad in action. When is the next release to be expected? Thanks

**Comment by canton7** [comment]
My bad, it's released now as 1.2.0. Thanks for your patience
